trio/_core/_io_windows.py

```diff
@@ -97,219 +222,280 @@ class CompletionKeyEventInfo:
-        IOCP_CLOSED_ERRORS = {
-            # If the IOCP is closed while we're blocked in
-            # GetQueuedCompletionStatusEx, then we get this error:
-            ErrorCodes.ERROR_ABANDONED_WAIT_0,
-            # If the IOCP is already closed when we initiate a
-            # GetQueuedCompletionStatusEx, then we get this error:
-            ErrorCodes.ERROR_INVALID_HANDLE,
-        }
-        while True:
-            max_events = 1
-            batch = ffi.new("OVERLAPPED_ENTRY[]", max_events)
-            received = ffi.new("PULONG")
-            # https://msdn.microsoft.com/en-us/library/windows/desktop/aa364988(v=vs.85).aspx
             try:
                 _check(
-                    kernel32.GetQueuedCompletionStatusEx(
-                        self._iocp, batch, max_events, received, 0xffffffff, 0
+                    kernel32.DeviceIoControl(
+                        self._afd,
+                        IoControlCodes.IOCTL_AFD_POLL,
+                        poll_info,
+                        ffi.sizeof("AFD_POLL_INFO"),
+                        poll_info,
+                        ffi.sizeof("AFD_POLL_INFO"),
+                        ffi.NULL,
+                        lpOverlapped,
                     )
                 )
             except OSError as exc:
-                if exc.winerror in IOCP_CLOSED_ERRORS:
-                    # The IOCP handle was closed; time to shut down.
-                    return
-                else:
-                    self._iocp_queue.append(exc)
-                    return
-            self._iocp_queue.append((batch, received[0]))
-            self._main_thread_waker.wakeup_thread_and_signal_safe()
+                if exc.winerror != ErrorCodes.ERROR_IO_PENDING:  # pragma: no cover
+                    raise
+
+            op = AFDPollOp(lpOverlapped, poll_info, waiters)
+            waiters.current_op = op
+            self._afd_ops[lpOverlapped] = op
+
+    async def _afd_poll(self, sock, mode):
+        base_handle = _get_base_socket(sock)
+        waiters = self._afd_waiters.get(base_handle)
+        if waiters is None:
+            waiters = AFDWaiters()
+            self._afd_waiters[base_handle] = waiters
+        if getattr(waiters, mode) is not None:
+            raise _core.BusyResourceError
+        setattr(waiters, mode, _core.current_task())
+        self._refresh_afd(base_handle)
+
+        def abort_fn(_):
+            setattr(waiters, mode, None)
+            self._refresh_afd(base_handle)
+            return _core.Abort.SUCCEEDED
+
+        await _core.wait_task_rescheduled(abort_fn)
 
     @_public
-    def current_iocp(self):
-        return int(ffi.cast("uintptr_t", self._iocp))
+    async def wait_readable(self, sock):
+        await self._afd_poll(sock, "read_task")
+
+    @_public
+    async def wait_writable(self, sock):
+        await self._afd_poll(sock, "write_task")
+
+    @_public
+    def notify_closing(self, handle):
+        handle = _get_base_socket(handle)
```

Comment:
I wonder if it's worth marking these `_get_base_socket` calls with some indication (a comment or a chained exception) that failure usually means the argument is not a socket. I figure that's the first thing that will go wrong if someone tries to use a library that does Unix-specific `wait_readable` calls on Windows, so it might pay to be user-friendly here.

Reply:
For `notify_closing` I think we want to eventually accept non-SOCKET handles: #1270.

You're right though that we should probably have a test that `wait_*` on non-SOCKETs gives a sensible result...

Reply:
It looks like the error we get currently is:

```
OSError: [WinError 10038] An operation was attempted on something that is not a socket
```

Which I guess is pretty clear on it not being a socket :-)

Reply:
great!